drivers: adc: stm32: Ensure ADRDY is awaited after enabling adc

abxy opened the pull request:
I encountered a problem using the ADC on STM32WL55 where the conversion completion interrupt would not fire. Using `git bisect` I found that there was a regression introduced by commit 58b8c4006e1bc0bb2ee8056b807ecb11e1be71bb.

I found that the ADC had to be disabled when configuring oversampling, and then re-enabled, otherwise some other bits in the ADC_CFGR2 register would reset. This seems to have been partially fixed by 97c031822bf023145e8b54492fc09b41df2fe719, only to regress again at acad37196d1f9bf32e82a8af6f085d6f8eb356aa.

Of all revision mentioned so far 97c031822bf023145e8b54492fc09b41df2fe719 comes the closest to working. I do get conversions but the value is always 0. The sample prints `ADC reading: 0`. Waiting for ADRDY when re-enabling the ADC fixes that, and since this is done at a couple other places I added a wrapper for `LL_ADC_Enable` which includes the wait on ADRDY. 

This PR also adds support for running the ADC sample on nucleo_wl55jc, which is what I've been using to test. If this is undesired I'll break it out into its own PR.

abxy committed:
boards: nucleo_wl55jc: Add support for ADC samples

Add pinmux for ADC on Arduino pin A0 and enable ADC1.

Signed-off-by: Alexander Mihajlovic <a@abxy.se>

abxy committed:
samples: adc: Add nucleo_wl55jc support

Add support for nucleo_wl55jc board to adc driver sample.

Signed-off-by: Alexander Mihajlovic <a@abxy.se>

abxy committed:
drivers: adc: stm32: Add function to enable ADC consistently

Use a wrapper for LL_ADC_Enable that also waits for ADRDY if required
by the SoC to make sure it's properly enabled everywhere this is done.

Signed-off-by: Alexander Mihajlovic <a@abxy.se>

erwango commented on samples/drivers/adc/boards/nucleo_wl55jc.overlay:
```diff
@@ -0,0 +1,12 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+
```
Fix empty line

abxy replied:
Done

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -402,6 +402,41 @@ static void adc_stm32_oversampling(ADC_TypeDef *adc, uint8_t ratio, uint32_t shi
+	 * still not stabilized, this will wait for a short time to ensure ADC
+	 * modules are properly enabled.
+	 */
+	uint32_t countTimeout = 0;
```
Let's take the opportunity to fix this to `count_timeout` in conformance with coding guidelines.

abxy replied:
Good point. Done.

erwango reviewed (changes requested):
Some minor comments, but thanks for this fix, which also adds some clarity to this driver.

erwango commented:
@abxy I forgot about on going https://github.com/zephyrproject-rtos/zephyr/pull/41376, can you have a check?
This should also fix the issue you were seeing.

This being said, I still see value in your PR, so what about rebasing it on top of #41376 once that one is merged ?

abxy commented:
@erwango I forgot to look for open PRs when investigating this. I'll rebase once #41376 is merged. Meanwhile I updated the commit message to better reflect the remaining changes.

abxy committed:
tests: drivers: adc: Add support for nucleo_wl55jc

Add support for nucleo_wl55jc board to adc_api test.

Signed-off-by: Alexander Mihajlovic <a@abxy.se>

abxy commented:
The remaining twister failures seem like test cases that are not meant to run for `nucleo_wl55jc`. Both of them fail on the main brainch as well. Try `twister -p nucleo_wl55jc -T tests/lib/devicetree`.

I can get both failing tests to succeed for `nucleo_wl55jc` adding the following to their `prj.conf` files:
```
CONFIG_GPIO=n
CONFIG_PINCTRL=n
CONFIG_SERIAL=n
```

The first line `CONFIG_GPIO=n` solves the initial problem of duplicate gpio device definitions. The other two lines are needed to solve build errors caused by excluding GPIO. A similar problem has been adressed before by 54fb930febee991037f7992f000293e93e48008b which describes the issue clearly.

Let me know if you want me to adress this in any way. I personally feel like its outside the scope of this PR.

FRASTM commented on drivers/adc/adc_stm32.c:
```diff
@@ -402,6 +402,41 @@ static void adc_stm32_oversampling(ADC_TypeDef *adc, uint8_t ratio, uint32_t shi
+		}
+	}
+#endif
+
```
should the ADC Ready bit ADRDY bit be cleared afterwards ?

abxy replied:
I just looked in RM0453. Under 18.3.4 ADC on-off control (ADEN, ADDIS, ADRDY) it says to clear ADRDY as _before_ enabling and (optionally) after _disabling_.

It's cleared by software so obviously it has to be cleared sometime or the check is pointless. Currently this is done in `adc_stm32_init` before enabling the ADC, but it's not apparent from the code if this is the intent.

abxy replied:
I had a look in a few other reference manuals and most of them agree that ADRDY should be cleared before enabling, RM0456 says to clear it afterwards as well, RM0316 doesn't mention clearing it although that seems like an oversight. Every manual that I checked says to wait for ADRDY after enabling, but this is not done for all relevant series at the moment.

I'm not sure how to proceed. My aim was to fix an issue that I was having with STM32WL. I believe this PR can serve as a first step toward improving the handling of ADRDY but I unfortunately don't have resources to fix it for every case.

erwango replied:
Are the tests available in zephyr sufficient to test this use case ?
If this is the case, I propose we test the current change on our board farm.

erwango replied:
@abxy, @FRASTM can you comment my proposal?

abxy replied:
@erwango What do you mean by "this use case"? Are you referring to the discussion above, or to the failure I have encountered? The sample `samples/drivers/adc` is sufficient to test the failure I'm having. Every sample read is 0.

```
** Booting Zephyr OS build v2.7.99-2866-g4963959afc12  ***
ADC reading: 0
ADC reading: 0
ADC reading: 0
```

Inspecting the `handler.log` for `tests/drivers/adc/adc_api` running on target it consistently reads a 0 for `test_adc_sample_one_channel` before my change, but this is not considered a failure. The other tests in the same suite don't read all 0's, but they all seem to be doing repeated samples in one way or another.

Command to run test on target:
```
twister --device-testing --device-serial /dev/ttyACM0 --device-serial-baud 115200 -p nucleo_wl55jc  -T tests/drivers/adc/adc_api/
```

erwango replied:
@abxy I meant the ADRDY handling, if we attempt to change it on all series (assuming we'll be able to test on all series)

abxy replied:
@FRASTM  I had a check in a bunch more reference manuals and I think the best approach is to clear ADRDY before enabling. I added a separate commit for this change.

@erwango What type of test can you run on your board farm? Every series affected by the ADRDY handling (L4, L5, WB, G0, G4, H7, WL) is covered by a Nucleo board configuration enabled in `tests/drivers/adc/adc_api`. Unfortunately the test itself passes a zero valued sample which means it won't catch all errors related to this.

The `dac_loopback` test would be better suited I think, but it doesn't seem to have coverage for all series. I'm trying to get it to run on my board to test if it can trigger my original problem.

erwango replied:
> What type of test can you run on your board farm?

All tests in tree tests. We also have (or will add if missing) fixtures for things like loopback tests.

> The dac_loopback test would be better suited I think, but it doesn't seem to have coverage for all series.

We can extend coverage as long as dac is available on these series.

abxy replied:
For some reason `tests/drivers/dac_loopback` doesn't trigger this issue for me. I don't understand exactly why. I can modify the test it to get the issue to trigger, but small changes can make it disappear again. The adc sample on the other hand shows the issue reliably even, when I add sleeps of varying length between calls to the adc api.

erwango commented:
> The remaining twister failures seem like test cases that are not meant to run for nucleo_wl55jc. Both of them fail on the main brainch as well. 

I just had a try and these errors should go away after a rebase (should have been fixed meanhile).

abxy committed:
drivers: adc: stm32: Clear ADRDY before waiting

Clear ADRDY before enabling ADC to ensure the subsequent
wait for ADRDY does not stop prematurely in case ADRDY
was already set.

The "ADC on-off control" sections of the following reference manuals
were consulted. That gives at least one instance per series affected
by this change, even if not every affected MCU is covered.

- RM0438 (STM32L552xx and STM32L562xx)
- RM0351 (STM32L47xxx, STM32L48xxx, STM32L49xxx and STM32L4Axxx)
- RM0434 (STM32WB55xx and STM32WB35xx)
- RM0454 (STM32G0x0)
- RM0440 (STM32G4 Series)
- RM0399 (STM32H745/755 and STM32H747/757)
- RM0433 (STM32H742, STM32H743/753 and STM32H750)
- RM0453 (STM32WL5x)

Signed-off-by: Alexander Mihajlovic <a@abxy.se>

erwango reviewed (commented):
@FRASTM Would you mind testing this on our bench ?

FRASTM commented:
connect pin A4 to A2 on CN8 of the nucleo_l55jc board
$ west build -p auto -b nucleo_wl55jc tests/drivers/dac/dac_loopback/

```
Running test suite dac_loopback_test
===================================================================
START - test_dac_loopback
 PASS - test_dac_loopback in 0.11 seconds
===================================================================
Test suite dac_loopback_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

$ west build -p auto -b nucleo_wl55jc tests/drivers/adc/adc_api
```
Running test suite adc_basic_test
===================================================================
START - test_adc_sample_one_channel
Samples read: 0x05aa 0xffff8000 0xffff8000 0xffff8000 0xffff8000 0 
 PASS - test_adc_sample_one_channel in 0.8 seconds
===================================================================
START - test_adc_sample_two_channels
 SKIP - test_adc_sample_two_channels in 0.1 seconds
===================================================================
START - test_adc_asynchronous_call
Samples read: 0x05c8 0x0593 0x057c 0x0573 0x056d 0xffff8000 
 PASS - test_adc_asynchronous_call in 0.6 seconds
===================================================================
START - test_adc_sample_with_interval
sample_with_interval_callback: sampling 0
sample_with_interval_callback: sampling 1
sample_with_interval_callback: sampling 2
sample_with_interval_callback: sampling 3                          
sample_with_interval_callback: sampling 4                          
Samples read: 0x05d6 0x05b6 0x05b1 0x05ad 0x05af 0xffff8000        
 PASS - test_adc_sample_with_interval in 0.410 seconds             
===================================================================
START - test_adc_repeated_samplings                                
repeated_samplings_callback: done 1                                
Samples read: 0x05c9 0xffff8000 0xffff8000 0xffff8000 0xffff8000 0 
repeated_samplings_callback: done 2                                
Samples read: 0x05c9 0x0574 0xffff8000 0xffff8000 0xffff8000 0xfff 
repeated_samplings_callback: done 3                                
Samples read: 0x05c9 0x0599 0xffff8000 0xffff8000 0xffff8000 0xfff 
repeated_samplings_callback: done 4                                
Samples read: 0x05c9 0x0558 0xffff8000 0xffff8000 0xffff8000 0xfff 
repeated_samplings_callback: done 5                                
Samples read: 0x05c9 0x0594 0xffff8000 0xffff8000 0xffff8000 0xfff 
repeated_samplings_callback: done 6                                
Samples read: 0x05c9 0x0553 0xffff8000 0xffff8000 0xffff8000 0xfff 
repeated_samplings_callback: done 7                                
Samples read: 0x05c9 0x0592 0xffff8000 0xffff8000 0xffff8000 0xfff 
repeated_samplings_callback: done 8                                
Samples read: 0x05c9 0x0551 0xffff8000 0xffff8000 0xffff8000 0xfff 
repeated_samplings_callback: done 9                                
Samples read: 0x05c9 0x0592 0xffff8000 0xffff8000 0xffff8000 0xfff 
repeated_samplings_callback: done 10                               
Samples read: 0x05c9 0x054f 0xffff8000 0xffff8000 0xffff8000 0xfff 
 PASS - test_adc_repeated_samplings in 0.97 seconds                
===================================================================
START - test_adc_invalid_request                                   
E: Invalid resolution                                              
E: Invalid resolution                                              
Samples read: 0x05ca 0xffff8000 0xffff8000 0xffff8000 0xffff8000 0 
 PASS - test_adc_invalid_request in 0.12 seconds                   
===================================================================
Test suite adc_basic_test succeeded                                
===================================================================
PROJECT EXECUTION SUCCESSFUL  
```

erwango commented:
@abxy Would you mind renaming this PR as this is no more about fixing a regression on WL (which could be confusing)

abxy commented:
@erwango Is the new title ok?

FRASTM commented:
When testing with the stm32G474 on the nucleo_g474re, a deadlock occurs appears. 
 The adc_stm32_enable is called twice 

1. during adc_stm32_init  --> the LL_ADC_Enable is correct, the ADC Ready bit is set, the LL_ADC_IsActiveFlag_ADRDY while-loop exits
2. during the start_read : the ADC ready flag (was set) is cleared, the ADC  is re-enabled , but the ADC ready flag remains low so the while-loop endless ( LL_ADC_IsEnabled is 1 so timeout does not decrease)

What could be done is to avoid the sequence if the ADC is already enabled.

FRASTM commented on drivers/adc/adc_stm32.c:
```diff
@@ -407,15 +407,17 @@ static void adc_stm32_oversampling(ADC_TypeDef *adc, uint8_t ratio, uint32_t shi
 	defined(CONFIG_SOC_SERIES_STM32H7X) || \
 	defined(CONFIG_SOC_SERIES_STM32WLX)
+
+	LL_ADC_ClearFlag_ADRDY(adc);
```
```
	LL_ADC_ClearFlag_ADRDY(adc);
	if (LL_ADC_IsEnabled(adc) == 1UL) {
		return 0;
	}
	LL_ADC_Enable(adc);
```

abxy replied:
I took your suggestion but put it before LL_ADC_ClearFlag_ADRDY as well because there is no point in clearing the flag if we're not gonna await it.

I tested this before and after on a nucleo_wb55rg which I just got in the mail. :) It also had the deadlock issue.
